sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/NuGetMetadataReference.cs

```diff
@@ -163,19 +166,41 @@ private static void EnsurePackageIsInstalled(string packageId, string packageVer
                 }
             }
         }
-
+        
+        /// <summary>
+        /// Returns the list of folders containing installed versions of the specified package,
+        /// or an empty list if the package is not installed.
+        /// </summary>
+        /// <remarks>
+        /// Package directory names are in the form "{package id}.{package version}".
+        /// The list is sorted in ascending order, so the most recent version will be last.
+        /// </remarks>
+        private static IEnumerable<string> GetSortedPackageFolders(string packageId)
+        {
+            // The package will be in a folder called "\packages\{packageId}.{version}", but:
+            // : the package might not be installed
+            // : there might be multiple versions installed
+            // : there might be a package that starts with the same package id
+            //      e.g. Microsoft.AspNetCore.Core and Microsoft.AspNetCore.Core.Diagnostics
+            // Most packages have a three-part version, but some have four. We don't check
+            // the actual number of parts, as long as there is at least one.
+            var matcher = new Regex($"{packageId}(.\\d+)+$");
+
+            var directories = Directory.GetDirectories(PackagesFolderRelativePath, $"{packageId}.*", SearchOption.TopDirectoryOnly)
+                .Where(path => matcher.IsMatch(path))
+                .OrderBy(name => name)
+                .ToArray();
+            return directories;
+        }
+        
         private static string GetLastCheckFilePath(string packageId)
         {
             // The file containing the last-check timestamp is stored in folder of the
             // latest version of the package.
-            // Package directory names are in the form "{package id}.{package version}".
-            // Sorting the names orders them by version.
-
             const string LastUpdateFileName = "LastCheckedForUpdate.txt";
 
-            var directory = Directory.GetDirectories(PackagesFolderRelativePath, $"{packageId}.*")
-                .OrderByDescending(name => name)
-                .FirstOrDefault();
+            var directory = GetSortedPackageFolders(packageId)
+                .LastOrDefault();
```

Comment:
FYI I was getting a test failure because I didn't have the _Microsoft.AspNetCore.2.1.6_ package installed. The test framework was doing a check to see if it needed to check for the latest version, and this method finding the _LastCheckedForUpdate.txt_ for one of the other AspNetCore packages e.g. (_Microsoft.AspNetCore.Diagnostics.2.1.1_) and deciding that it didn't need to check for the latest package -> exception because the package was not installed.